### PR TITLE
Codecov plugin

### PR DESCRIPTION
**Pull request by zofpolkowska**
Extract codecov-compatible coverage report generating code to separate Erlang project and use as a plugin for small tests

### CODE REVIEW

**Comment by mongoose-im** [comment]
[5516.1](https://travis-ci.org/esl/MongooseIM/jobs/423872325) / Erlang 19.3 / small_tests / d15db4cc72bce800614e043bc66f090da7962dfa
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2059/5516/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2059/5516/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-03_11.41.22/lib.mongooseim.logs/run.2018-09-03_11.41.23/suite.log.html)

**Comment by mongoose-im** [comment]
[5520.1](https://travis-ci.org/esl/MongooseIM/jobs/423917250) / Erlang 19.3 / small_tests / 91b330dc714ca6e1f6353c9575ae8a474b57b288
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2059/5520/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2059/5520/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-03_14.07.00/lib.mongooseim.logs/run.2018-09-03_14.07.00/suite.log.html)

File: rebar.config [hunk 1]
```diff
@@ -138,12 +138,14 @@
...
-  {provider_asn1, {git, "https://github.com/knusbaum/provider_asn1.git", {ref, "29f7850"}}}
+  {provider_asn1, {git, "https://github.com/knusbaum/provider_asn1.git", {ref, "29f7850"}}},
+  {rebar3_codecov, {git, "https://github.com/zofpolkowska/rebar3_codecov.git", {branch, "WIP"}}}
  ]}.
 
 {provider_hooks,
  [{pre,  [{compile, {asn, compile}}, {compile, {pc, compile}}]},
-  {post, [{clean, {asn, clean}}, {clean, {pc, clean}}]
+  {post, [{clean, {asn, clean}}, {clean, {pc, clean}}],
+   {post, [{cover, {codecov, analyze}}]}
```
Comment (arcusfelis) [inline rebar.config]: This solution ignores `COVER_ENABLED` env variable.

(unless it's handled somewhere ine plugin itself).

**Comment by mongoose-im** [comment]
[5528.1](https://travis-ci.org/esl/MongooseIM/jobs/424268038) / Erlang 19.3 / small_tests / 2495107e38e567b2af5ca70bf380f1a6aec45613
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2059/5528/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2059/5528/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-04_10.43.28/lib.mongooseim.logs/run.2018-09-04_10.43.28/suite.log.html)

**Comment by mongoose-im** [comment]
[5531.6](https://travis-ci.org/esl/MongooseIM/jobs/424308416) / Erlang 19.3 / elasticsearch_and_cassandra_mnesia / f58e041c7416da97c2ef653a2dde2933711ab044
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2059/5531/elasticsearch_and_cassandra_mnesia.19.3)/ [big](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2059/5531/elasticsearch_and_cassandra_mnesia.19.3/big/ct_run.test%40travis-job-b9a1568d-1bec-47e4-85e2-5ded755591bd.2018-09-04_12.49.29/index.html)
**OK: 445** / Failed: 0 / **User-skipped: 8** / Auto-skipped: 0

File: rebar.config [hunk 2]
```diff
@@ -138,12 +138,14 @@
...
-  {provider_asn1, {git, "https://github.com/knusbaum/provider_asn1.git", {ref, "29f7850"}}}
+  {provider_asn1, {git, "https://github.com/knusbaum/provider_asn1.git", {ref, "29f7850"}}},
+  {rebar3_codecov, {git, "https://github.com/zofpolkowska/rebar3_codecov.git", {branch, "master"}}}
  ]}.
 
 {provider_hooks,
  [{pre,  [{compile, {asn, compile}}, {compile, {pc, compile}}]},
-  {post, [{clean, {asn, clean}}, {clean, {pc, clean}}]
+  {post, [{clean, {asn, clean}}, {clean, {pc, clean}}],
+   {post, [{ct, {codecov, analyze}}]}
```
Comment (michalwski) [inline rebar.config]: We only need codecov analyze when tests are run on Travis. It's not necessary when we test locally. With this change the analysis will be executed every time.

File: rebar.config [hunk 3]
```diff
@@ -138,12 +138,14 @@
       {ref, "eb3cded5b050edd82cf8653f8c850c6c9890f732"}}},
   {pc, {git, "https://github.com/blt/port_compiler.git", {ref, "c2f3fb1"}}},
   {coveralls, {git, "https://github.com/michalwski/coveralls-erl.git", {ref, "f124a62"}}},
-  {provider_asn1, {git, "https://github.com/knusbaum/provider_asn1.git", {ref, "29f7850"}}}
+  {provider_asn1, {git, "https://github.com/knusbaum/provider_asn1.git", {ref, "29f7850"}}},
+  {rebar3_codecov, {git, "https://github.com/zofpolkowska/rebar3_codecov.git", {branch, "master"}}}
```
Comment (michalwski) [inline rebar.config]: I suggest to freeze the version when the plugin is finished. Rebar plugins doesn't go to rebar.lock so currently the newest code will be pulled when rebar needs to download the plugin.

**Comment by mongoose-im** [comment]
[5539.1](https://travis-ci.org/esl/MongooseIM/jobs/425142569) / Erlang 19.3 / small_tests / 9cbe2f77a26289e56ccd24834d420a3b61c8452d
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2059/5539/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2059/5539/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-06_07.33.08/lib.mongooseim.logs/run.2018-09-06_07.33.08/suite.log.html)

**Comment by mongoose-im** [comment]
[5547.1](https://travis-ci.org/esl/MongooseIM/jobs/425240014) / Erlang 19.3 / small_tests / 4775d1e7e2e080189ba234516ff78162a427b057
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2059/5547/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2059/5547/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-06_13.50.06/lib.mongooseim.logs/run.2018-09-06_13.50.06/suite.log.html)

**Comment by mongoose-im** [comment]
[5548.1](https://travis-ci.org/esl/MongooseIM/jobs/425606302) / Erlang 19.3 / small_tests / 291d00c3296c41e7e66447fc8bdadf7b0163921e
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2059/5548/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2059/5548/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-07_07.11.47/lib.mongooseim.logs/run.2018-09-07_07.11.48/suite.log.html)

File: tools/travis-test.sh [hunk 4]
```diff
@@ -75,6 +75,7 @@ run_small_tests() {
   SMALL_SUMMARIES_DIRS=${BASE}/_build/test/logs/ct_run*
   SMALL_SUMMARIES_DIR=$(choose_newest_directory ${SMALL_SUMMARIES_DIRS})
   ${TOOLS}/summarise-ct-results ${SMALL_SUMMARIES_DIR}
+  DEBUG=1 ./rebar3 codecov analyze
```
Comment (michalwski) [inline tools/travis-test.sh]: I think, it'd be better to call this in `.travis.yml` just before sending the report to codecov.

Comment (michalwski) [reply tools/travis-test.sh]: or at least under the `$COVER_ENABLED` condition as in the removed `run_small_tests_cover_report` function.

Comment (zofpolkowska) [reply tools/travis-test.sh]: I'd go with the first option

**Comment by mongoose-im** [comment]
[5551.1](https://travis-ci.org/esl/MongooseIM/jobs/425738421) / Erlang 19.3 / small_tests / daeaf904c2b3deb441d68407cb1c735a3352c4cb
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2059/5551/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2059/5551/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-07_13.51.05/lib.mongooseim.logs/run.2018-09-07_13.51.06/suite.log.html)

**Comment by mongoose-im** [comment]
[5562.1](https://travis-ci.org/esl/MongooseIM/jobs/426621906) / Erlang 19.3 / small_tests / d573a039b9f6e83c9c0c96bbf6cb3b44ce7b7756
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2059/5562/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2059/5562/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-10_11.02.40/lib.mongooseim.logs/run.2018-09-10_11.02.40/suite.log.html)

**Comment by mongoose-im** [comment]
[5563.1](https://travis-ci.org/esl/MongooseIM/jobs/426625008) / Erlang 19.3 / small_tests / f56282d6ee801e622b37c1f4ada59a9fd680f4db
Reports [root](http://esl.github.io/mongooseim-ct-reports/s3_reports.html?prefix=PR/2059/5563/small_tests.19.3) / [small](https://mongooseim-ct-results.s3-eu-west-1.amazonaws.com/PR/2059/5563/small_tests.19.3/small/ct_run.mongooseim@localhost.2018-09-10_11.49.17/lib.mongooseim.logs/run.2018-09-10_11.49.18/suite.log.html)

**Comment by michalwski** [comment]
To be 100% it works I suggest to close this PR and open new. One check from codecov failed and I'm not sure what's the root cause.

**Pull request closed without merging**